pkgs/sops-install-secrets/main.go

```diff
@@ -509,6 +510,74 @@ func importSSHKeys(keyPaths []string, gpgHome string) error {
 	return nil
 }
 
+func handleRestart(symlinkPath string, secretDir string, secrets []secret) error {
+	var restart []string
+	var reload []string
+
+	// When /run/booted-system does not exist yet, we are being run in stage-2-init.sh
+	// where switch-to-configuration is not run so the services would only be restarted
+	// the next time switch-to-configuration is run.
+	if _, err := os.Stat("/run/booted-system"); os.IsNotExist(err) {
+		return nil
+	}
+	for _, secret := range secrets {
+		// Skip when there are no restarts/reloads
+		if len(secret.RestartUnits) == 0 && len(secret.ReloadUnits) == 0 {
+			continue
+		}
+		oldPath := filepath.Join(symlinkPath, secret.Name)
+		newPath := filepath.Join(secretDir, secret.Name)
+
+		// Read the old file
+		oldData, err := ioutil.ReadFile(oldPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// File did not exist before
+				restart = append(restart, secret.RestartUnits...)
+				reload = append(reload, secret.ReloadUnits...)
+				continue
+			}
+			return err
+		}
+
+		// Read the new file
+		newData, err := ioutil.ReadFile(newPath)
+		if err != nil {
+			return err
+		}
+
+		if !bytes.Equal(oldData, newData) {
+			restart = append(restart, secret.RestartUnits...)
+			reload = append(reload, secret.ReloadUnits...)
+		}
+	}
+
+	// Write the services to restart/reload
+	writeLines := func(list []string, file string) error {
+		if len(list) != 0 {
+			f, err := os.OpenFile(file, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			for _, unit := range list {
+				if _, err = f.WriteString(unit + "\n"); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	}
+	if err := writeLines(restart, "/run/systemd/restart-list"); err != nil {
```

Comment:
Maybe just read both files restart/reload and everything not in the list can be restarted/reloaded from sops-nix itself.
Advantages: This works without changing nixpkgs. If nixpkgs's interface ever changes we might just restart/reload too often, which is not that much of a big deal and gives us some time to fix it.